### ISSUE TITLE
bug in favourites sorting

### ISSUE THREAD

**Issue by usemore** [xbmc/xbmc#23936]
kodi android master and others exhibit the same bug in favo(u)rites sorting...
if you move an item down from the bottom of the list, it used to go to the top.
now it crashes kodi.

if you move it up (painful!!!) all the way to the top, it just vanishes.

**Comment by ksooo** [comment]
Cannot reproduce. No crash, no vanishing items. 

Checked both favorites window and Estuary favorites widget. 

Definitely more infos needed, that’s what the issue template is for.

**Comment by usemore** [comment]
i cant imagine explaining it any better. been going on in all versions of 21 on android.
i have tested on 4 android devices, same result every time.

**Comment by ksooo** [comment]
Then you won’t get help, sorry. I have no crystal ball to guess what exactly you have done. All I can say is that it works here for me. 

 You are not able to explain what you did, which version exactly, which skin? Then you are lost.

**Comment by usemore** [comment]
in fact, it's worse now...moving my most used items to the top line of favourites just makes them vanish completely. i didnt do anything out of the ordinary. been using kodi/xbmc for 20 years

**Comment by usemore** [comment]
i have not installed any skins. using Estuary.

**Comment by ksooo** [comment]
Fill the issue template, provide all requested information.

**Comment by vpeter4** [comment]
I can reproduce this easily with LibreELEC. If @usemore will fail to provide the required information then I will give exact procedure and debug log.

**Comment by ksooo** [comment]
@vpeter4 best for the crash is a full stack trace.

**Comment by usemore** [comment]
vpeter4, i would appreciate you submitting the debug info for this.. i opened the issue because i wanted a place to track the bug and i didnt see it out there.. btw, i dont use librelec just kodi on 4 versions of android, fireos7, and googletv

**Comment by ksooo** [comment]
@usemore what the hell is your problem with giving us the information we need in a structured way, by filling out the issue form?

**Comment by usemore** [comment]
i'm not a developer, i'm a longtime fan of kodi and power-end-user i guess.. i'm very new to github, and while i've tried multiple times to get useful logs from kodi i never managed to.
if someone wants to give me clear concise instructions on generating a trace or log that would be useful to them, i will be happy to assist but i also dislike forms..i went back to edit it and couldnt find an 'edit button'. i was honestly hoping someone like vpeter4 would pipe up.

btw, i still use kodi 14 on an appletv2, it's working well as a headless upnp/dlna renderer... and i'm still using XBMC on an OG XBOX!

**Comment by ksooo** [comment]
You really need not to be a developer to fill the issue template give us step-by-step instructions what you did to run into the crash. This is pure user point of view. No developer skills needed for this at all.

Regarding debug log. There are so many postings out there explaining how to do this. Ask Google. ;-)

I'm out here now. Maybe somebody else will pick this up...

**Comment by usemore** [comment]
honestly, the crash was common across several platforms so i just figured i'd point it out and get the ball rolling

**Comment by vpeter4** [comment]
The result is not always the same but I'm providing 4 log files.
It started with 4 items in favorites. After moving last (forth) one down it disappear in the list and first item is duplicated and not always crash happen. Then moving next last (third) item down crash happen.
Content of `favourites.xml `:
```
<favourites>
    <favourite name="Item 3" thumb="�^"></favourite>
    <favourite name="Item 1" thumb="image://pvrchannel_tv@%2fstorage%2fitem.png/">PlayMedia("pvr://channels/tv/All%20channels@-1/1@pvr.iptvsimple_29511301.pvr")</favourite>
    <favourite name="Item 1" thumb="image://pvrchannel_tv@%2fstorage%2fitem.png/">PlayMedia("pvr://channels/tv/All%20channels@-1/1@pvr.iptvsimple_29511301.pvr")</favourite>
    <favourite name="Item 2" thumb="image://pvrchannel_tv@%2fstorage%2fitem.png/">PlayMedia("pvr://channels/tv/All%20channels@-1/1@pvr.iptvsimple_29511302.pvr")</favourite>
</favourites>
```

three screenshots:
![screenshot1_initial](https://github.com/xbmc/xbmc/assets/1016324/b69745ad-f35e-4b15-ba23-d648f61bc133)
![screenshot2_item4_move_down](https://github.com/xbmc/xbmc/assets/1016324/cbdf61c7-c97d-4122-9fee-3a720b080200)
![screenshot3_item3_move_down](https://github.com/xbmc/xbmc/assets/1016324/5f6c99c6-516d-4634-b150-bfeaa14d4c2b)

four log files:
[log_files.zip](https://github.com/xbmc/xbmc/files/12910081/log_files.zip)

**Comment by ksooo** [comment]
What is this? Broken xml file? Definitely looks not like it should.

```
<favourite name="Item 3" thumb="�^"></favourite>
```

**Comment by vpeter4** [comment]
Yes, it get broken when last item is moved second time.

**Comment by ksooo** [comment]
Tried hard, but cannot reproduce. I guess the crash is just the result of something that goes wrong before. No idea what this could be.

**Comment by ksooo** [comment]
Could be a multithreading issue. The code is not thread safe, but could be the case it gets called from multiple threads concurrently.

**Comment by ksooo** [comment]
Yes, definitely can be called from multiple threads, no wonder we see such weird behavior for you, but not for others, like me. It's all timing-dependent.

**Comment by vpeter4** [comment]
For me crash happens sometimes on first move but on second or third move I get 100% crash.
And this is done on fresh kodi profile with nothing set just 4 favorite items. Kodi probably doesn't do much in this case.

**Comment by ksooo** [comment]
@vpeter4 in caes you can build yourself, could you apply this patch and report back?

```diff
diff --git a/xbmc/favourites/FavouritesService.cpp b/xbmc/favourites/FavouritesService.cpp
index 73b034c229..5eaed1cddb 100644
--- a/xbmc/favourites/FavouritesService.cpp
+++ b/xbmc/favourites/FavouritesService.cpp
@@ -147,6 +147,8 @@ CFavouritesService::CFavouritesService(std::string userDataFolder) : m_favourite
 
 void CFavouritesService::ReInit(std::string userDataFolder)
 {
+  std::unique_lock<CCriticalSection> lock(m_criticalSection);
+
   m_userDataFolder = std::move(userDataFolder);
   m_favourites.Clear();
   m_targets.clear();
@@ -288,6 +290,8 @@ std::shared_ptr<CFileItem> CFavouritesService::ResolveFavourite(const CFileItem&
 {
   if (item.IsFavourite())
   {
+    std::unique_lock<CCriticalSection> lock(m_criticalSection);
+
     const auto it = m_targets.find(item.GetPath());
     if (it != m_targets.end())
       return (*it).second;
```

**Comment by KarellenX** [comment]
There does seem to be something strange happening with favourites.
It seems that when an item is moved to the top, it is rewritten in the favorites.xml as the item that is second in the list.
See video...

https://github.com/xbmc/xbmc/assets/44626605/386cd48c-df1c-4308-a616-2ab2bd6a3b4d

**Comment by ksooo** [comment]
@vpeter4 in case you are here and able/willing to build/test, here is another patch you could add in addition to the first one, then report back:

```diff
diff --git a/xbmc/favourites/FavouritesUtils.cpp b/xbmc/favourites/FavouritesUtils.cpp
index 59fffc110a..3efcf5ca2c 100644
--- a/xbmc/favourites/FavouritesUtils.cpp
+++ b/xbmc/favourites/FavouritesUtils.cpp
@@ -84,18 +84,16 @@ bool MoveItem(CFileItemList& items, const std::shared_ptr<CFileItem>& item, int
   if (itemPos < 0 || itemPos >= items.Size())
     return false;
 
-  int nextItem = (itemPos + amount) % items.Size();
+  const int nextItem{(itemPos + amount) % items.Size()};
   if (nextItem < 0)
   {
-    const auto itemToAdd(item);
-    items.Remove(itemPos);
-    items.Add(itemToAdd);
+    items.Add(item);
+    items.Remove(0);
   }
   else if (nextItem == 0)
   {
-    const auto itemToAdd(item);
-    items.Remove(itemPos);
-    items.AddFront(itemToAdd, 0);
+    items.AddFront(item, 0);
+    items.Remove(itemPos + 1);
   }
   else
   {
```

**Comment by vpeter4** [comment]
First patch didn't made any difference.
But second one fixed the issue. I did repeat the previous tests and moving items works as expected.

**Comment by ksooo** [comment]
> But second one fixed the issue. I did repeat the previous tests and moving items works as expected.

@vpeter4 thanks for reporting back. I will PR the changes the next days.
